### ISSUE TITLE
Version 4 - Design Documents

### ISSUE THREAD

**Issue by xelatihy** [xelatihy/yocto-gl#1151]
This issue tracks design consideration for the v4 simplification. Instead of keeping separate design documents for each part of the library, we keep everything into one for simplicity.

**Comment by xelatihy** [comment]
# Color Images

There have been two problems we tackled in v3 and before. The main problem is that different image formats makes sense for different applications, but we do not want to support that as discussed in other issues. Instead we want to cut down code size while still allowing to support everything we do now. Here are the things a few things that we care about:

1. keep float and byte images
2. perform all computation in float, exception for image sampling which can be done in bytes
3.  we seldomly save data in byte for speed
4.  we compute in linear and non-linear spaces depending on the application
5.  makes sense to keep color computation in vec3f, but store always in vec4f or vec4b 

So this means we have 4 variants

1. float / linear (HDR)
2. float / non-linear (computed values)
3. byte / linear (normal maps)
4. byte / nonlinear (albedo maps)

A major issue seems to be that we started the project thinking byte=non-linear and 
float=linear and that was a big mistake. That lead to weird things everywhere in the pipeline,
typeing to perform conversions in places where none was desired. We have to try to fix it now.
A major issues that needs to be addressed is _loading should never change color space,
and saving should always respect it_.

Given all these constraints, and the discussion below, the proposed implementation is

- [x] implement a shared type for float and bytes, for now with only 4 channels
- [x] stored a flag for llinear/non-linear (in leu of proper color spaces)
- [x] access pixels _only_ with set_pixel()/get_pixel()
- [x] add functions width and height
- [x] make images by default as non-linear floats
- [x] load and save images by respecting the color space
- [x] sample images within their color space, but with a flag to force linear
- [x] resize images in their natural representation
- [x] tone map from linear to non-linear, default as float, but with a construction flag for bytes
- [x] color grade from linear to non-linear, default to float, but with a construction flag for bytes
- [x] possibly wrap this into a class wo we get proper behavior and add color spaces later
- [x] implement set/get region

**Follows a discussion on design choices**

Here are some operations we perform and how that relates to image formats.

1. tone mapping: HDR to LDR, but more importantly linear to non-linear clamped
2. color grading: HDR to LDR or LDR to LDR, i.e. linear/non-linear to non-linear clamped
3. loading: either float-only or float/byte based on filename
    - should loading ignore color space or marked images with it?
4. texturing: all the above, but generally all converted to linear depending on use
5. resizing: same color space and flaot/byte type
6. diffing: same color space and float/byte type
7. viewing: all cases, although we could skip the float/byte conversion in the view API

The question now to decide how to handle all these cases in code.

1. One option is to have a templated type as we have now
    - library is easy to but apps are painful to write
    - many duplications on types
2. The other option is to have a single type with both byte/float and a linear flag
    - loading: always as is and keep color space; can force to float
    - saving: float converted appropriately based on color space and file formats, bytes as they are
    - resizing: as is
    - tone map: linear to non-linear; input only float; output byte or float, but default to float
    - make proc: can choose to output floats, but mark linear/non-linear
    - colorgrade: input as float, output as float; linear/non-linear to non-linear
    - eval: float/byte input, float output
    - bytes images are used little, but maybe helpful at times: textures (so eval), eval, output for tonemap/grade only for speed

**Comment by xelatihy** [comment]
# Generic Shapes

In v3, Yocto did not have a concept of shape explicit within the library. Only low-level functions were used throughout. That makes it possible to build custom apps very easily. But this is at the price of usability and simplicity. In this version, we are planning to introduce two common data types for shapes that are either indexed or face-varying. The discussion here will mostly focus on the indexed shapes, since the face-varying ones are nearly identical.

The main features we want to support are:

1. generic shape data that is, for now, not extensible to arbitrary data
2. methods for building shapes with attributes
3. query methods for shape types and properties (no more `!triangles.empty()` idiom)
4. vertex data getter and setter, to be used optionally
5. vertex data interpolation for the built it properties (these were in the path tracer before)
6. loading and saving shapes
7. high level helpers: smoothing normals and targets, subdivision, etc

It seems that to do these, there is very little design that still needs doing, besides naming conventions that are now not properly designed now. So we can just make a list of things to implement.

- [ ] generic shape data
- [ ] initialization functions for shapes
- [ ] query functions for shape types and vertex property
- [ ] vertex and face setter/getter
- [ ] vertex interpolating functions with eval_XXX methods
- [ ] loading and saving shapes, possibly with forced convertions
- [ ] shape convertions placeholders (quads/triangles, but maybe also lines/points to surfaces)
- [ ] normals smoothing and faceting
- [ ] subdivision

For subdivision, things are very similar, so we just add things here.

- [ ] generic face-varying shape data
- [ ] initialization functions for face-varying shapes
- [ ] query functions for face-varying shape types and vertex property
- [ ] vertex and face setter/getter
- [ ] vertex interpolating functions with eval_XXX methods
- [ ] loading and saving face-varying shapes
- [ ] shape convertions placeholders (face-varying to indexed and vice-versa)
- [ ] normals smoothing and faceting
- [ ] subdivision

**Comment by xelatihy** [comment]
# Scene creation

We have gone back and forth on scene creation, mostly to handle the complexity of either too many parameters, the pointer usage and so on. Right now, we are settling on a simple design that is natural to a data-oriented C++ stack. This design preserves the following goals:

1. scene creation should be natural to avoid having to write tons of getters and setters
2. data is accessed directly, so we should do so when we can
3. in one alternative, scene creation is just manipulation of the various arrays in the scene
4. add_XXX functions helpful to build a scene in a programmatic apron since they are potentially shorter to use and allow for presets; they also hide the complexity of the materials, at least for now
5.  names are optional and do not have to be set; add methods will take them always for simplicity
6. add methods should for now take only simple params, even if they eventually will take the full description
7. add methods should be the exact same in the glscene

Given this design, the translation to todos is easy

- [ ]  remove all add_XXX function from scene that do nothing
- [ ] add add_XXX function that create full objects
- [ ] possibly add make_XXX functions that create full objects, useful for cameras, materials, environments and instances
- [ ] add_XXX functions should add names when necessary
- [ ] handle name creation in IO and presets
- [ ] eventually move all presets to the loading of images, shapes and materials; but for presets we should probably decid about how to expose these to the outside
- [ ] update the glscene to have the same methods, although for now it may be required to have some clutches for untyped materials and shapes

**Comment by xelatihy** [comment]
# Names in IO

This is a placeholder for now for a design discussion on the use of names.

**Comment by xelatihy** [comment]
# Library organization

This is a placeholder now for library organization.

**Comment by xelatihy** [comment]
# Simpler Material

The material model of v3 and before is derived from the Disney model. That has worked well so far, but has two main drawbacks: complexity in creation and complexity in rendering. In the years this has been hard to use for students, which should not have happened. We are now proposing to following pbrt and use a tagged material model. We want the following features:

1. materials should have intuitive types types
2. use types instead of flags to change major beavior
3. if possible, match pbrt materials in the definitions
4. convert materials to/from other formats: maintain compatibility with old format and focus on glTF for forward compatibility
5. reduce the use of textures significantly and if possible to two textures
6. consider whether to set colors as srgb or rub linear --- it is possible than srgb is simpler to handle overall
7. in the renderer, use switches to handle brdf parameters

Given these goals, the first issue to solve is to deign the material types and parameters. We would base it on the current versions and then normalize it after.

From this discussion, we now need the following parameters:

- matte: color, opacity, color_tex
- plastic: color, opacity, color_tex, for, roughness, specular_tex
- metal: color, opacity, color_tex, roughness, specular_tex
- metallic: color, opacity, color_tex, metallic, roughness, specular_tex
- thin glass: color, opacity ior, roughness, specular_tex 
- thick glass: color, opacity, ior, roughness, specular_tex, scattering, phaseg
- subsurface: color, opacity, ior, roughness, specular_tex, scattering, phaseg
- volume: only volumetric param

Additional property might include:

- light: emission, emission_tex
- normal map, normal_tex

Given these goals, the following changes are needed:

- [ ] review pbrt materials
- [ ] define new material parametrization
- [ ] save old parametrization for file compatibillity
- [ ] change renderer to halde these
- [ ] change glrender to handle them, probably with no materials types for now, but doing a simpler translation: use metallic as the main type

**Comment by xelatihy** [comment]
# Rationalization of IO

Right now Yocto is mostly independent since many of its IO needs are served within Yocto libraries. This has served us very well in the past. But it is beginning to take a toll in our ability to move forward. All these IO needs require significant effort and a lot of maintenance, that it is unlikely to scale well in the future.

In the end, our goals are

- make it easy to write Yocto apps, command line and gui
- keep the IO the same through all yocto libraries
- reduce maintenance costs

These goals are well served with this refactoring

- [ ] command line
    - [ ] simplify calls
    - [ ] remove command/cli_state difference
    - [ ] consider using CLI11 internally
    - [ ] consider using macros to autogenerate cli commands
    - [ ] consider using labels for enums
- [ ] remove Yocto/JSON
- [ ] gui
    - [ ] add a generic parameter structure
    - [ ] consider macros to go backs and forth
    - [ ] otherwise consider serialization methods
- [ ] modelio
    - [ ] consider wrapper for STL
    - [ ] evaluate more carefully OBJ
    - [ ] evaluate more carefully PLY
    - [ ] evaluate more carefully PBRT

**Comment by xelatihy** [comment]
# OpenGL Objects

This is a placeholder for discussions about OpenGL in the library.

**Comment by xelatihy** [comment]
# Large tests

With all changes to Yocto/GL so far, we should probably start to rerun all tests. This will require selecting tests, updating the tests scripts and providing rendering configurations for all.

To run tests we first need to update some infrastructure of the various scripts we are using.

- [ ] remove logo
    - [ ] remove logo
    - [ ] update scene script
- [ ] params from files
    - consider if doing this in the script for now

**Comment by xelatihy** [comment]
# Apps

This is a placeholder for discussions about Apps in the library.

**Comment by xelatihy** [comment]
# Path tracing

Placeholder for improvements to the path tracer in v4.

**Comment by xelatihy** [comment]
# Material Naming

This issue tracks a discussion of materials naming and parameters.

There are four main concerns:
- material names
- what to do with emission 
- what to do with eta/ior
- what to do with volumetric properties

# Material Types

The current implementation supports the following materials:
- matte: Lambert brdf with `color` as diffuse
- metal: specular brdf with `color` as reflectivity at grazing angles (conductor fresnel for now) and `roughness`
- plastic: labert + specular with `color` as diffuse, `ior` and `roughness` for specular
- thinglass: thin transmission and reflection with `color` for transmission color and `ior` and `roughness` for specular/fresnel
- glass: same as before but with refraction and homogenous volume; volume uses `color` and `depth` for density and `scattering` and `anisotropy` for albedo
- metallic: glTF material with diffuse + specular, uses `color`, `metallic` and `roughness`
- subsurface: same as glass above with `scattering` texture and to allow a different way of entering the surface; it is ignored for now but I think it is better to keep a subsurface material for skin rendering
- volume: a purely volumetric material with bo surface interface, for now uses only the volumetric properties

# Material naming

Material naming is another complex issues. All renderers have different kind of nomenclature, so there is no common ground to use. There are three main variations for this

- names of common materials: matte, plastic, glass
    -  used in pbrt-v3
    - looks fine to me but gets confusing when we have glass for water or leaves from diffuse translucency
- names from physics: dielectric, conductor, coateddiffuse
    - usd in tungsten, pbrt-v4
    - to me it is not good
    - it implies that we support correctly physics behaviors 
    - it implies that you need to know physics to understand this, namely it scares people away
- names from phenomenological appearance: diffuse, reflective (metal), transmissive, refractive, translucent
   - used by nobody, but the one I prefer

I prefer the last naming convention.

# Material emission

Right now emission is kept as a separate property, just like in all renderers. This may or may not be a good idea.

- Pros: Simpler IO
- Pros: Other do it this way
- Pros: `color` is a color in [0,1], while `emission` is in [0,inf]; we could fix this by adding an `exposure` param
- Cons: can be confusing for new people looking for "lights"
- Cons: need more setup since we should not use a black BRDF by default, maybe
- Cons: 1 more param, 1 more texture
- Cons: cannot simulate funky behaviors, like coated emission (glow stick)
- Cons: less obvious where lights are (very minor)

Here I am not sure what to do.

# Material parameters

The following parameters are r mostly easonable, besides possibly coming up with better names.

- `color`: this is the main object reflectance, but it really is different in all materials; generally this is the one that gets the texture
    - called sometimes `reflectance` in pbrt-v4
    - I would probably leave it like this
    - `color_tex` for texture, but could be `color_map`
- `opacity`: clear what it is, I would leave it like this
    - alpha from `color_tex` 
- `roughness`: clear what it is, I would leave it like this
    - `roughness_tex` for texture, but could be `roughness_map`
- `metallic`: used only for glTF, I would leave it as is since it matches those specs
    - texture in `roughness_tex`
- `emission`: see above
    - `emission_tex` as texture
- `normal_tex` for normal maps; works like it ism but maybe call it `normal_map`

The more controversial parameters are `ior` and all volumetric settings.

For metals, we can use either Schlick's or conductors' Fresnel terms. I refer to introduce all the complexity of complex for frankly since we have no data and since it cannot be textured. So a simple `reflectivity` setting, as a color, works well enough.

For dielectrics it is harder. Right now we use `ior` for two things: coating/plastic and transmission/refraction. For coating, a reflectivity settings works fine and in fact is required in glTF. Before we used to have `specular` for that, but it was used in awkward ways, so it was changed to single `ior`. For transmission/refraction, `ior` is used in dielectric fresnel and refraction direction, but has two issues: (1) we do not do spectral rendering, so we need single value `ior`, and (2) nobody changes it since who knowns what it does.

WE can leave this as single `ior`, basically a read-only property. for now, I use dielectric Fresnel, but it is impossible to teach and explain (in fact all renders have the same cut&pasted code). I would not have a 3 channel `ior` since it is completely strange (see pbrt-v4 scenes). Th other option is to use `specular` set to {0.04, 0.04, 0.04} and be done with it (gt for when needed). In the future, that can be used for other effects, like colored coatings.

Not sure what to do here.

# Vometric properties

The other hard issue are volumetric properties. These are nasty since they are not intuitive no matter what one does. For this we have three materials now:
- glass: meant for refractive glass
- subsurface: meant for skin
- volume: meant for smoke and clouds, i.e. no interface

In the end, we need to set absorption, albedo and anisotropic (phase g), but it is helpful to consider the there cases separately.
 - glass: typically absorption only, so make sense that color control this, maybe
 - subsurface: need two set absorption but main color controls scattering, for example by using the Sony Parametrization to go from diffuse color to scattering
 - volume: need to control both absorption and scattering: main color from scattering, I am guessing (colored smoke?)
 
This already would suggest different parametrizations for at least glass and subsurface. Or maybe it make sense to keep the same one, which is what is done in production (glass: color -> density, subsurface: mfp -> density).

BTW, the renderer should also be changed based on these materials a bit. For example if no scattering is there, a simpler early exit is useful, but is for another day.

Here I am really not sure what do. Noe I know what to do with volumetric textures since I have not tried any data yet.

The bottom line is that we could follow two way:

- current paraterization:
    - color -> density
    - scattering -> albedo, probably named `subsurface`
- different parametrization for glass and subsurface where color sets different properties
    - glass though would remain as is now

I am not sure what to do.

**Comment by xelatihy** [comment]
# Prepare for Release

This is a list of final things we need to do to get the release ready.

- [ ] fix all tests
    - [ ] decide whether to automatically generate tests 
- [ ] resolve the scene names issues
- [ ] resolve material names
- [ ] resolve material parametrization for volumes
- [ ] resolve skin rendering

Here are the things that can be postponed.

- [ ] path tracing improvements listed above
- [ ] path tracing API cleanup

**Comment by xelatihy** [comment]
The release is ready.
